### PR TITLE
[Fix #9803] Update Bundler/GemVersion to check commit references

### LINKED ISSUE #9803: Bundler/GemVersion not respecting tags
## Expected behavior

I expect the `Bundler/GemVersion` to respect tags on private repositories as "correct", since they lock the version.
This was working fine in the previous versions

## Actual behavior

The Cop is currently flagging all private gems as "wrong"

```
Gemfile:78:3: E: Bundler/GemVersion: Gem version specification is required.
  gem 'inkycop', github: 'customink/inkycop', tag: '3.4.1'
  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

```
# Base gem for all Custom Ink Clients, pinned to this version to handle compatibility issues.
gem 'client_client', github: 'customink/client_client', tag: 'v0.2.1'
# Client to obtain design information from Custom Ink
gem 'design_client', github: 'customink/design_client', tag: 'v1.4.5'
# GDPR Implementation for Ruby on Rails
gem 'gdpr_rails', github: 'customink/gdpr_rails', tag: '0.7.0'
# Gem to obtain information about ink-colors in designs
gem 'ink_colors', '>=1.7.4', git: 'git@github.com:customink/ink_colors.git', branch: 'master'
# Gem to determine whether the application is up and running
gem 'is_it_up', github: 'customink/is_it_up', tag: '1.0.0'
# Gem to communicate with Custom Ink MMS
gem 'mms_client', github: 'customink/mms_client', tag: 'v2.2.1'
# Gem to communicate with Custom Ink OFS
gem 'order_fulfillment_client', github: 'customink/order_fulfillment_client', tag: 'v0.4.12'
# Gem to communicate with Custom Ink OIS
gem 'order_infrastructure', git: 'git@github.com:customink/order_infrastructure.git', tag: 'v0.31.0'
# Gem to communicate with Custom Ink Mint Service
gem 'payment_client', github: 'customink/payment_client', tag: 'v1.4.0'
# Gem to communicate with Custom Ink Quote Service
gem 'quote_client', github: 'customink/quote_client', tag: 'v0.0.9'
```
## Steps to reproduce the problem

- Create a Gemfile with github references to repo's
- Set the version through tags
- Run RuboCop

## RuboCop version

Include the output of `rubocop -V` or `bundle exec rubocop -V` if using Bundler.
If you see extension cop versions (e.g. `rubocop-performance`, `rubocop-rspec`, and others)
output by `rubocop -V`, include them as well. Here's an example:

```
➜  customink_international git:(coding-bunny/bump_inkycop) ✗ bundle exec rubocop -V
1.14.0 (using Parser 3.0.1.1, rubocop-ast 1.5.0, running on ruby 3.0.0 x86_64-linux)
  - rubocop-performance 1.11.3
  - rubocop-rails 2.10.1
  - rubocop-rake 0.5.1
  - rubocop-rspec 2.3.0

```

### PR DESCRIPTION
**Pull request by timlkelly**
https://github.com/rubocop/rubocop/pull/9807 seems to be stale. This pull request completes the work it set out to do and will close #9803 and #9807.

I originally wanted to create a pull request to merge into https://github.com/rubocop/rubocop/pull/9807, however, I was not able to figure that out. Probably due to my inexperience when working on git projects across multiple forks.

This differs slightly from https://github.com/rubocop/rubocop/pull/9807 given the conversation that has happened in that pull request:
- This proposal does not take the hosted git service into consideration
- This proposal also checks for `branch`

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: changelog/fix_gem_version_not_respecting_tags.md [hunk 1]
```diff
@@ -0,0 +1 @@
+* [#9803](https://github.com/rubocop/rubocop/pull/9803): Fix `Bundler/GemVersion` cop not respecting git tags. ([@tejasbubane,@timlkelly][])
```
Comment (timlkelly) [inline changelog/fix_gem_version_not_respecting_tags.md:1]: I am not sure if the syntax is correct here for multiple contributors but I wanted to make sure to include both of our work here. Should this be different?

Comment (dvandersluis) [reply changelog/fix_gem_version_not_respecting_tags.md:1]: Each contributor would have a separate link. This should work:

```suggestion
* [#9803](https://github.com/rubocop/rubocop/pull/9803): Fix `Bundler/GemVersion` cop not respecting git tags. ([@tejasbubane][], [@timlkelly][])
```

Comment (timlkelly) [reply changelog/fix_gem_version_not_respecting_tags.md]: Ah, got it. Makes sense, thank you!

**Comment by bbatsov** [comment]
Looks good to me! Thanks for driving it to the finish line!

**Pull request merged**